homeassistant/components/spotify/__init__.py

```diff
@@ -53,39 +54,38 @@ async def async_setup_entry(hass: HomeAssistant, entry: SpotifyConfigEntry) -> b
     except aiohttp.ClientError as err:
         raise ConfigEntryNotReady from err
 
-    spotify = Spotify(auth=session.token["access_token"])
+    spotify = SpotifyClient(async_get_clientsession(hass))
 
-    coordinator = SpotifyCoordinator(hass, spotify, session)
+    spotify.authenticate(session.token[CONF_ACCESS_TOKEN])
 
-    await coordinator.async_config_entry_first_refresh()
+    refresh_lock = asyncio.Lock()
 
-    async def _update_devices() -> list[dict[str, Any]]:
-        if not session.valid_token:
+    async def _refresh_token() -> str:
+        async with refresh_lock:
             await session.async_ensure_token_valid()
-            await hass.async_add_executor_job(
-                spotify.set_auth, session.token["access_token"]
-            )
+            token = session.token[CONF_ACCESS_TOKEN]
+            if TYPE_CHECKING:
+                assert isinstance(token, str)
+            return token
```

Comment:
This is kind of stupid, why don't we make `async_ensure_token_valid` return the token, then this would just be simplified to:
`return session.async_ensure_token_valid()`

(For a different PR though)